rfcs/0154-Migrate-queue-to-postgres.md

```diff
@@ -0,0 +1,102 @@
+# RFC 154 - Migrate Taskcluster to postgres
+* Updates: RFC [65](https://github.com/taskcluster/taskcluster-rfcs/blob/master/rfcs/0065-Migrate-queue-to-postgres.md)
+* Obsoletes: None
+* Comments: [#154](https://github.com/taskcluster/taskcluster-rfcs/pull/154)
+* Initially Proposed by: @jonasfj
+
+# Summary
+
+Taskcluster uses Azure as the data-storage backend. For example, the Queue
+service currently uses a bunch of Azure tables to handle tasks, dependencies,
+artifacts, and so on. This project involves refactoring Taskcluster to use a
+Postgres backend, instead.
+
+# Motivation
+
+* A shared base of indexed data can allow services to make smarter decisions and
+  save money on provisioning and data storage
+* Azure is unreliable with frequent failures and inscrutible rate-limiting. While we are resilient
+to these errors, they do cause inefficiency and sometimes cascading failures
+* Storing data in postgres will enable features users desire:
+  * Queue introspection
+    * What tasks are pending right now?
+    * How many tasks are [actually pending and not cancelled](https://bugzilla.mozilla.org/show_bug.cgi?id=1434851)?
+  * Integration of task-related and worker-related status
+    * What tasks has this worker performed?
+  * Ad-hoc analytics
+    * How much compute time do we use per push?
+    * What do we pay for wpt?
+  * More sophisticated dependency resolution [in the case of task failure](https://bugzilla.mozilla.org/show_bug.cgi?id=1443503#c8)
+  * Dynamically changing task levels
+
+# Details
+
+## Transition
```

Comment:
I have run out of time to comment on this today, but I think this transition plan is currently much too risky. I'll followup with an explanation of why and thoughts on how we can change that.

Reply:
In order to switch data storage systems, we need to build a lot of confidence in the correctness and performance of our implementation of the new one. Planned approaches to doing this are worth including in the RFC.

The ideal solution would probably be to read and write from both data stores in parallel, compare the results and flag an error if they differ, then throw away the results from postgres. The amount of requests to mirror to postgres could be configurable, so we could easily turn it off if there were issues that affected taskcluster's availability during testing.

We should also think about if a big-bang cutover is absolutely necessary. Hopefully it is possible to do it more incrementally. In either case, we also need to think about how we would switch back from postgres to azure if needed.

Reply:
> we need to build a lot of confidence in the correctness and performance of our implementation of the new one. Planned approaches to doing this are worth including in the RFC.

Agreed. We can talk more about in our meeting next week. For now, we can get some confidence in the correctness and performance of our implementation by first testing the implementation in the development and staging environments before running this on the production development.